File: Array/901.Online_Stock_Span/C++/901.Online_Stock_Span.cpp
class StockSpanner 
{
    /*Initialize*/
    ///// Record stack /////
    stack<pair<int, int>> record_stack;

public:
    StockSpanner() 
    {
        //======================//
        // Input type:          //
        //   - void (no input)  //
        // Return type:         //
        //   - void (no return) //
        //======================//
        ;
    }

    ~ StockSpanner()
    {
        //======================//
        // Input type:          //
        //   - void (no input)  //
        // Return type:         //
        //   - void (no return) //
        //======================//
        ;
    }
    
    int next(int price) 
    {
        //=======================//
        // Input type:           //
        //   - int (price value) //
        // Return type:          //
        //   - int (result span) //
        //=======================//

        /*Initialize*/
        ///// Record span /////
        int record_span = 1;

        /*Whole process, flow*/
        while ((! (this->record_stack.empty())) && (price >= ((this->record_stack).top()).first)) //Part
        {
            record_span += ((this->record_stack).top()).second; //Keep updating, accumulating

            (this->record_stack).pop(); //Keep updating, popped

        } //Part

        (this->record_stack).emplace(price, record_span); //Keep updating, recording

        return record_span;
    }
};

/**
 * Your StockSpanner object will be instantiated and called as such:
 * StockSpanner* obj = new StockSpanner();
 * int param_1 = obj->next(price);
 */

File: Array/901.Online_Stock_Span/Python/901.Online_Stock_Span.py
class StockSpanner(object):
    def __init__(self):
        """
        :type: None, void
        :rtype: None, void
        """
        ############
        #Initialize
        ##### Record stack #####
        self.record_stack = []


    def next(self, price):
        """
        :type price: int
        :rtype: int
        """
        ############
        #Initialize
        ##### Record span #####
        record_span = 1

        ####################
        #Whole process/flow
        while ((self.record_stack) and ((self.record_stack[(-1)])[0] <= price)):
            record_span += ((self.record_stack).pop())[1] #Keep updating/accumulating

        (self.record_stack).append((price, record_span)) #Keep updating/recording

        return record_span
        


# Your StockSpanner object will be instantiated and called as such:
# obj = StockSpanner()
# param_1 = obj.next(price)
